File: algorithms/08_r/techniques/sliding_window.R
max_sum_subarray <- function(arr, k) {
  if (k <= 0 || length(arr) < k) return(0)
  window_sum <- sum(arr[1:k])
  max_sum <- window_sum
  n <- length(arr)
  if (k < n) {
    for (i in (k + 1):n) {
      window_sum <- window_sum + arr[i] - arr[i - k]
      max_sum <- max(max_sum, window_sum)
    }
  }
  max_sum
}

# Example
a <- c(2, 1, 5, 1, 3, 2)
stopifnot(max_sum_subarray(a, 3) == 9)
print("[R sliding_window] ... verified")

File: algorithms/05_csharp/techniques/sliding_window.cs
using System;

namespace Algorithms.Techniques
{
    public static class SlidingWindow
    {
        public static int MaxSumSubarray(int[] array, int k)
        {
            if (k <= 0 || array.Length < k) return 0;
            int windowSum = 0;
            for (int i = 0; i < k; i++) windowSum += array[i];
            int maxSum = windowSum;
            for (int i = k; i < array.Length; i++)
            {
                windowSum += array[i] - array[i - k];
                maxSum = Math.Max(maxSum, windowSum);
            }
            return maxSum;
        }

        public static void Main()
        {
            int[] a = { 2, 1, 5, 1, 3, 2 };
            bool ok = MaxSumSubarray(a, 3) == 9;
            Console.WriteLine(ok ? "[C# SlidingWindow] ... verified" : "[C# SlidingWindow] ... FAILED");
        }
    }
}
